### PR TITLE
libc: minimal: Add strtoll() and strtoull()

### LINKED ISSUE #36645: minimal libc: add strtoll and strtoull functions
**Is your enhancement proposal related to a problem? Please describe.**
I'm using a Nordic NRF9160 chip. This supports AT_PARSER a lib which servers the built-in cellular modem using AT commands. Due to that fact that this lib needs strtoll and strtoull they added a dependency to newlibc. This dependency is new with the latest SDK that has a new version of AT_PARSER. Using newlibc is not possible for me due to incompatibility with Civetweb.

**Describe the solution you'd like**
I would be happy if somebody could add strtoll and strtoull functions to the minimal libc.

**Describe alternatives you've considered**
I tried to switch to newlibc which didn't succeed due to a  incompatibility with Civetweb.

**Additional context**
[Discussion in Nordic Forum](https://devzone.nordicsemi.com/f/nordic-q-a/76717/nordic-nrf-sdk-v1-6-issue-with-at_cmd_parser
)
https://github.com/zephyrproject-rtos/zephyr/pull/16511

### PR DESCRIPTION
**Pull request by MeisterBob**
strtoll() and strtoull() are copies of strtol() and strtoul() with types
changed to long long instead of long.

Signed-off-by: Gerhard Jörges <joerges@metratec.com>

Closes #36645

### CODE REVIEW

**Review by tejlmand — DISMISSED** [review]

**Review by stephanosio — CHANGES_REQUESTED** [review]
1. Please add the tests for these functions in `tests/lib/c_lib`.
2. Need to update the relevant documentation:
https://github.com/zephyrproject-rtos/zephyr/blame/main/doc/guides/portability/posix.rst#L311
3. Need to remove the local `strtoll` implementation:
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/samples/net/civetweb/common/src/libc_extensions.c#L156
https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/samples/net/civetweb/common/include/libc_extensions.h#L29

**Comment by MeisterBob** [comment]
I was busy in an other project the last weeks, but finally I added tests and documentation and removed stubs from civetweb samples. Also rebased the branch..

**Comment by MeisterBob** [comment]
I stumbled across https://github.com/zephyrproject-rtos/zephyr/pull/16511 by @PiotrZierhoffer. I liek the way he unified the conversion mechanism. The only thing missing in his PR where tests. I would suggest to take the tests I implemented and put them into his PR.

File: tests/lib/c_lib/src/main.c [hunk 1]
```diff
@@ -812,6 +812,145 @@ void test_strtoul(void)
...
+	zassert_equal(ret, -1011, "strtoull base = 10 failed");
+	zassert_true((strcmp(stop, " This stopped it") == 0), "strtoull get stop failed");
+
+	ret = strtoull(str_abnormal, &stop, 0);
+	zassert_equal(ret, 0, "strtoull base = 0 failed");
+	zassert_true((strcmp(stop, "ABCDEFGH") == 0), "strtoull get stop failed");
+
+	char border1[] = "+18446744073709551615";
+	char border2[] = "-18446744073709551615000";
+	char border3[] = "+18446744073709551619";
```
Comment (rivimey) [inline tests/lib/c_lib/src/main.c:944]: Suggest adding a couple more cases here to test a long string of digits but not a large number, and a simply very large number:
```
border[] = "0x0000000000000000000000000000000000001"; // should pass
border[] = "10000000000000000000000000000000000001"; // should fail (too big)
border[] = "-10000000000000000000000000000000000001"; // should fail (too big)
```

Comment (MeisterBob) [reply tests/lib/c_lib/src/main.c:953]: added those cases to strtoll and strtoull test.

**Review by tejlmand — APPROVED** [review]
lgtm - reapprove

**Review by stephanosio — CHANGES_REQUESTED** [review]
This PR contains the source files that are `BSD-4-Clause-UC` licensed.

This is different to the usual 3-clause BSD license variants that we have previously accepted into the codebase; in particular, the third clause is likely going to be a problem for the project.

```
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgment:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
```

If we are to accept this as is, this needs to go through the board as per the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#components-using-other-licenses).

The most practical solution would be to replace the strto(u)ll implementation with the one that is more freely licensed (e.g. BSD-2-Clause, BSD-3-Clause, ...).

cc @nashif @carlescufi

**Comment by MeisterBob** [comment]
It's the same license [strtol](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtol.c) and  [strtoul](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtoul.c) are using. If you can point out an implementation that more suitable licensed I would be happy to replace strtol, strtoul, strtoll and strtoull.

**Comment by stephanosio** [comment]
> It's the same license [strtol](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtol.c) and [strtoul](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/libc/minimal/source/stdlib/strtoul.c) are using. If you can point out an implementation that more suitable licensed I would be happy to replace strtol, strtoul, strtoll and strtoull.

@MeisterBob You are right. They are indeed `BSD-4-Clause-UC` (not sure why these did not turn up while grepping ...).

In that case, I assume this license has already been approved by the board. I would like @nashif to give a confirmation though, since those date back to the beginning of the project (maybe before such rules were in place?) ...

**Review by carlescufi — CHANGES_REQUESTED** [review]
I am lookin at adding a few more of those. Can you please move those to a single source file and combine the use of the whitespace trimming so that it is not duplicated for every function?

**Review by carlescufi — APPROVED** [review]
>I am lookin at adding a few more of those. Can you please move those to a single source file and combine the use of the whitespace trimming so that it is not duplicated for every function?

Please ignore this, I will take care of this in a follow-up PR.

**Review by stephanosio — APPROVED** [review]

**Comment by de-nordic** [comment]
Shouldn't this addition be included into release notes?

**Comment by carlescufi** [comment]
This replaces #43289 for the mesh code, with further additions that I will provide.

**Review by de-nordic — APPROVED** [review]

**Comment by carlescufi** [comment]
> Shouldn't this addition be included into release notes?

Release notes are populated at release time generally, except for API breakage/changes.

**Comment by carlescufi** [comment]
Approved by the TSC, merging with failing checks:

- Coding guidelines: Use of reserved identifiers (this is the C library)
- Scancode: Use of a non-Apache license (approved by the TSC)

**Comment by kestewart** [comment]
This is a non-OSI approved license.  Per:https://docs.zephyrproject.org/latest/contribute/external.html#external-src-process it's going to need board review.

**Pull request merged**